site/source/docs/compiling/Deploying-Pages.rst

```diff
@@ -0,0 +1,162 @@
+=====================
+
+To provide the best possible user experience, make sure that the different ways that the page can fail are taken into account, and good error reporting is provided to the user. In particular, proceed through the following checklist for best practices.
+
+- Aim to fail as early as possible. A large source of frustration for users comes from scenarios where user's system is not ready to run the given page, but the error only becomes apparent after having waited for a minute to download 100MB worth of assets. For example, try to allocate the needed heap memory up front before actually loading up the page. This way if the memory allocation fails, the failure is immediate and no asset downloads need to be attempted at all.
+
+- If a particular browser is known to not be supported, resist the temptation to read ``navigator.userAgent`` field to gate users with that browser, if at all possible. For example, if your page needs WebGL 2 but Safari is known not to support it, do not exclude out Safari users with a following type of check:
+
+   .. code:: js
+
+    if (navigator.userAgent.indexOf('Safari') != -1) alert('Your browser does not support WebGL 2!');
+
+but instead, detect the actual errors:
+
+   .. code:: js
+
+    if (!canvas.getContext('webgl2')) alert('Your browser does not support WebGL 2!'); // And look for webglcontextcreationerror here for an error reason.
+
+This way the page will be future compatible once support for the particular feature later becomes available.
+
+- Test various failure cases up front by simulating different issues and browser limitations. For example, on Firefox, it is possible to manually disable WebGL 2 by navigating to ``about:config`` and setting the preference ``webgl.enable-webgl2`` to ``false``. This allows you to debug what kind of error reporting your page will present to the user in such a scenario. To disable WebGL support altogether for testing purposes, set the preference ``webgl.disabled`` to ``true``.
+
+- When working with IndexedDB, prepare to handle out of quota errors when user is about to run out of free disk space or allowed quota for the domain.
+
+- Simulate out of memory errors by allocating unrealistically much memory for ``WebAssembly.Memory`` object and for the preloaded file packages, if using any. Make sure that out of memory errors are flagged correctly as such (and reported to the user or to an error database).
+
+- Simulate download timeouts either intrusively by programmatically aborting XHR downloads, physically disconnecting network access, or by using external tools such as Fiddler. These types of tools can show up a lot of unexpected failure cases and help diagnose that the error handling path for such scenarios is as desired.
+
+- When developing the page locally, perform testing by using a local web server and not just via ``file://`` URLs. The script ``emrun.py`` in Emscripten source tree is designed to serve as an ad hoc web server for this purpose.
```

Comment:
Emrun is much superior to python SimpleHTTPServer, and is practically built on the same code as SimpleHTTPServer, but with proper mime type handling, `Large-Allocation` header and `Content-Encoding: gzip` support. Instead of calling `python -m SimpleHTTPServer 8000`, one types `./emrun --no_browser --port 8000 .` or `./emrun ./page.html` if one want to launch the file directly. I'll add a comment about that.